keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,518 @@
+1. Local KEK is used for encryption based on policy (N events, X time, etc)
+
+Since key hierarchy is implemented at the KMS plugin level, it should be seamless for the kube-apiserver. So whether the plugin is using a key hierarchy or not, the kube-apiserver should behave the same.
+
+What is required of the kube-apiserver is to be able to tell the KMS plugin which KEK (local KEK or KMS KEK) it should use to decrypt the incoming DEK. To do so, upon encryption, the KMS plugin could provide the encrypted local KEK as part of the `metadata` field in the `EncryptResponse`. The kube-apiserver would then store it in etcd next to the DEK. Upon decryption, the kube-apiserver provides the encrypted local KEK in `metadata` and `observedKeyID` from the last encryption when calling Decrypt. In case no encrypted local KEK is provided in the `metadata`, then we can assume key hierarchy is not used. The KMS plugin would query the external KMS to use the remote KEK to decrypt the DEK (same behavior as today). No state coordination is required between different instances of the KMS plugin.
+
+For the reference KMS plugin, the encrypted local KEK is stored in etcd via the `metadata` field, and once decrypted, it can be stored in memory as part of the KMS plugin cache to be used for encryption and decryption of DEKs. The encrypted local KEK is used as the key and the decrypted local KEK is stored as the value.
+
+```proto
+message EncryptResponse {
+    // The encrypted data.
+    bytes cipher = 1;
+    // The KMS key ID used for encryption operations.
+    // This can be used to drive rotation.
+    string currentKeyID = 2;
+    // Additional metadata to be stored with the encrypted data.
+    // This metadata can contain the encrypted local KEK that was used to encrypt the DEK.
+    // Stored unencrypted in etcd.
+    map<string, string> metadata = 3;
+}
+```
+
+The `DecryptRequest` passes the same `currentKeyID` and `metadata` returned by the previous `EncryptResponse` of this data as its `observedKeyID` and `metadata` for the decryption request.
+
+```proto
+message DecryptRequest {
+    // Version of the KMS plugin API.
+    string version = 1;
+    // The data to be decrypted.
+    bytes cipher = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
+    // The keyID that was provided to the apiserver during encryption.
+    // This represents the KMS KEK that was used to encrypt the data.
+    string observedKeyID = 4;
+    // Additional metadata that was sent by the KMS plugin during encryption.
+    map<string, string> metadata = 5;
+}
+
+message DecryptResponse {
+    // The decrypted data.
+    bytes plain = 1;
+    // The KMS key ID used to decrypt the data.
+    string currentKeyID = 2;
+    // Additional metadata that was sent by the KMS plugin.
+    map<string, string> metadata = 3;
+}
+
+message EncryptRequest {
+    // Version of the KMS plugin API.
+    string version = 1;
```

Comment:
I think we had decided to drop all the `version` bits from all of the APIs except `StatusResponse`?